src/kOS.Safe/Execution/CPU.cs

```diff
@@ -413,11 +415,13 @@ public void AssertValidDelegateCall(IUserDelegate userDelegate)
         {
             if (userDelegate.ProgContext != currentContext)
             {
-                throw new KOSInvalidDelegateContextException(
-                    (currentContext == contexts[0] ? "the interpreter" : "a program"),
-                    (currentContext == contexts[0] ? "a program" : "the interpreter")
-                    );
-            }
+                string currentContextName = (currentContext == contexts[0] ? "the interpreter" : "a program");
+                string delegateContextName = (
+                    userDelegate.ProgContext == contexts[0] ? currentContextName = "the interpreter" : (
+                        currentContext == contexts[0] ? "a program" : "a different program from a previous run" ) );
+
+                throw new KOSInvalidDelegateContextException(currentContextName, delegateContextName);
+           }
```

Comment:
Your re-write altered the logic.

Possible valid phrasings for the exception message are:

* Cannot call this lock or function from ``the interpreter`` when it was declared in ``a program``.
* Cannot call this lock or function from ``a program`` when it was declared in ``the interpreter``.
* Cannot call this lock or function from ``a program`` when it was declared in ``a different program from a previous run``.

What your first condition ends up printing the same string for both, which makes the nonsensical message:

* Cannot call this lock or function from ``the interpreter`` when it was declared in ``the interpreter``.

The only reason for the ugly additional message (the change I made from the red text you see above in github), is that previously it was making the false assumption that at least one of the two contexts must be the interpreter, therefore if one is a program than the other must be the interpreter.  This is false because they can be two different programs. (but can't be two different interpreters).

The truth table is this:

```
 Is currentContext | is userDelegate.ProgramContext ||  Resulting message
   equal to        |   equal to contexts[0] ?       ||
  contexts[0] ?    |                                ||
 ================  | ============================   || ===========================
                   |                                ||  n/a  (This case is impossible because
     yes           |        yes                     ||  this code is inside an if-check that only
                   |                                ||  is true when the two contexts are different.)
 ----------------  | ---------------------------    ||  ------------------------
                   |                                ||  currentContextName = "the interpreter"
     yes           |        no                      ||  delegateContextName = "a program"
                   |                                ||
 ----------------  | ---------------------------    ||  ------------------------
                   |                                ||  currentContextName = "a program"
     no            |        yes                     ||  delegateContextName = "the interpreter"
                   |                                ||
 ----------------  | ---------------------------    ||  ------------------------
                   |                                ||  currentContextName = "a program"
     no            |        no                      ||  delegateContextName = "a different program from a previous run"
                   |                                ||
```
Re-write that if/else logic however you like as long as it matches that truth table like the original expression did.  

The plain English version of the logic is basically "say 'interpreter' for the one that's context[0], and 'program' for the one that isn't, but also check for the case where they're both 'program' but different programs, and make sure the phrasing makes sense in that edge case".